contracts/multisig-prover/src/execute.rs

```diff
@@ -126,31 +128,123 @@ pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractEr
+                &cur_worker_set,
+                config.worker_set_diff_threshold as usize,
+            ) {
+                return Err(ContractError::WorkerSetUnchanged);
+            }
+
+            NEXT_WORKER_SET.save(
+                deps.storage,
+                &(new_worker_set.clone(), workers_info.snapshot),
+            )?;
+            let mut builder = CommandBatchBuilder::new(config.destination_chain_id);
+            builder.add_new_worker_set(new_worker_set)?;
+
+            let batch = builder.build()?;
+
+            COMMANDS_BATCH.save(deps.storage, &batch.id, &batch)?;
+            REPLY_BATCH.save(deps.storage, &batch.id)?;
+
+            let start_sig_msg = multisig::msg::ExecuteMsg::StartSigningSession {
+                key_id: cur_worker_set.id(), // TODO remove the key_id
+                msg: batch.msg_to_sign(),
+            };
+
+            Ok(Response::new().add_submessage(SubMsg::reply_on_success(
+                wasm_execute(config.multisig, &start_sig_msg, vec![])?,
+                START_MULTISIG_REPLY_ID,
+            )))
+        }
     }
+}
 
-    NEXT_WORKER_SET.save(deps.storage, &new_worker_set)?;
-    let mut builder = CommandBatchBuilder::new(config.destination_chain_id);
-    builder.add_new_worker_set(new_worker_set)?;
+pub fn confirm_worker_set(deps: DepsMut) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
 
-    let batch = builder.build()?;
+    let (worker_set, snapshot) = NEXT_WORKER_SET.load(deps.storage)?;
 
-    let start_sig_msg = multisig::msg::ExecuteMsg::StartSigningSession {
-        key_id: "static".to_string(), // TODO remove the key_id
-        msg: batch.msg_to_sign(),
+    let query = voting_verifier::msg::QueryMsg::IsWorkerSetConfirmed {
+        new_operators: worker_set.clone().into(),
+    };
+
+    let is_confirmed: bool = deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+        contract_addr: config.voting_verifier.to_string(),
+        msg: to_binary(&query)?,
+    }))?;
+
+    if !is_confirmed {
+        return Err(ContractError::WorkerSetNotConfirmed);
+    }
+
+    CURRENT_WORKER_SET.save(deps.storage, &worker_set)?;
+    NEXT_WORKER_SET.remove(deps.storage);
+    KEY_ID.save(deps.storage, &worker_set.hash().to_hex())?;
+
+    let key_gen_msg = multisig::msg::ExecuteMsg::KeyGen {
+        key_id: worker_set.id(),
+        snapshot, // TODO: refactor this to just pass the WorkerSet struct
+        pub_keys_by_address: worker_set
+            .signers
+            .into_iter()
+            .map(|s| {
+                (
+                    s.address.to_string(),
+                    (KeyType::Ecdsa, s.pub_key.as_ref().into()),
+                )
+            })
+            .collect(),
     };
 
-    // TODO handle the reply
-    Ok(Response::new().add_message(wasm_execute(config.multisig, &start_sig_msg, vec![])?))
+    Ok(Response::new().add_message(wasm_execute(config.multisig, &key_gen_msg, vec![])?))
```

Comment:
why remove the todo?

Reply:
this is not a submessage so it doesn't get a reply